### ISSUE TITLE
Discussion : support "sub-experiences"

### ISSUE THREAD

**Issue by Andlat** [G-Lauz/amnesis#8]
**See PR #5 for context.**

Should we allow reuse of a same ExperimentContext, and if so, how should it be logged ? 
Right now, #5 creates a new context for each call to the decorated function. 
We can also create new contexts manually, and re-use a model while only changing the data or hyperparameters. This logs it into Amnesis as a new experiment.

**Sooooooo**

- Do we want to allow "sub" experiments where the model and the hyperparameters are kept, but only the data is modified ? **--** Useful for benchmarking
- Does changing hyperparameters count as a separate or a sub experience ? **--** can be useful for parameter searching (e.g. PID factors)<br/>*Does this overlap with other tools like Optuna ?*

Other concerns ?

**Comment by G-Lauz** [comment]
Taking inspiration from [wandb](https://wandb.ai/site) and [trackio](https://huggingface.co/blog/trackio) we could implement something like:
```Python
with ExperimentContext(model_name="my_model") as ctx:
    for dataset in datasets:
        ctx.init()
        metric = my_cool_experiment(dataset)
        experiment.log_metric("my_metric", metric)
        ctx.finish()
```
Where the `init` and `finish` method could be part of a context manager, such as:
```Python
with ExperimentContext(model_name="my_model") as ctx:
    for dataset in datasets:
        with Experiment(ctx):
            metric = my_cool_experiment(dataset)
            experiment.log_metric("my_metric", metric)
```
But I'm not sure if I like the several indentations from this pattern.

**Comment by G-Lauz** [comment]
BRAINSTORM:

---

tldr:
- why not allowing (almost) unlimited sub-experience tree structure
- maybe this feature already exist and we must rename some methods and classes to make it more explicit

---

In facts, we can see this feature as a way of organizing experiments like we do with a tree of directories. Each experiments with different hyperparameters or dataset should create a new version under the same model name. So with this in mind, and currently as it is implemented the model name is the equivalent of the root directory, the version is a file in that directory. Suggesting a "sub-experience" means that we would have one more directory level (a sub-version). If we allow that, why not allow an (almost) unlimited sub-experiences?

I think the concepts of sub-experiences might already exist as is it is implemented and maybe we only need to rename some class to make this more explicit. As example:
```Python
# create as much version as dataset in DATASETS under the name model1
for dataset in DATASETS:
    with ExperimentContext(model_name="model1") as ctx:
        alpha = 42
        metric = my_experiment(alpha, dataset)
        ctx.log_hyperparameter("alpha", alpha)
        ctx.log_metric("metric", metric)

# create as much version as dataset in DATASETS under the name model2 (with different architecture/hyperparameters)
for dataset in DATASETS:
    with ExperimentContext(model_name="model2") as ctx:
        alpha = 0
        metric = my_experiment(alpha, dataset)
        ctx.log_hyperparameter("alpha", alpha)
        ctx.log_metric("metric", metric)
```

**Comment by Andlat** [comment]
As you said, this currently works by creating a different uuid/hash and a separate directory in .amnesis.
#5 was about bringing this same behavior to the @remember decorator

If we change the amnesis directory structure to be based on models for top-level directories, should we create uuid for the model specifically ? Something like :

```
uuid_model1
    hash_sub_exp_1
    hash_sub_exp_2
uuid_model2
    ...
```

Do this break the definition of "experiment" ?
Let's say we iterate on training an AI with different hyperparameters, would you consider it a "sub-experiment" or a different one ?
Do we consider the model to be only the architecture or the architecture + weights/params ?

If we just use the architecture as the model, we might encounter the issue of "extracting" the best weights to use it for inference.
We'll definitely have to revisit the display/listing of experiments

*We could also have both options* and an arg in the ExperimentContext constructor to indicate if it might be a sub-exp based on the model/hyperparams.



→ I was also wondering if we want to allow different hyperparameters between sub-experiments. I'm not really familiar with Optuna, but isn't specifically made for hyperparm research ?

**Comment by Andlat** [comment]
> Taking inspiration from [wandb](https://wandb.ai/site) and [trackio](https://huggingface.co/blog/trackio) we could implement something like:
> 
> with ExperimentContext(model_name="my_model") as ctx:
>     for dataset in datasets:
>         ctx.init()
>         metric = my_cool_experiment(dataset)
>         experiment.log_metric("my_metric", metric)
>         ctx.finish()

I was also thinking of reusing the experiment context to create sub-experiments... We could just have a `ctx.init_sub_exp()` that creates and logs metrics in a sub-experiment when called. And it's scope lasts until the context manager exists or `ctx.init_sub_exp()` is recalled.

```python
with ExperimentContext(model_name="model") as ctx:
    alpha = 0
    ctx.log_hyperparameter("alpha", alpha)

    for dataset in DATASETS:
        ctx.init_sub_exp() # something like that
        
        metric = my_experiment(alpha, dataset)
        ctx.log_metric("metric", metric)
```
We could have hyperparameters common to all sub-experiments, or set different values after `ctx.ini_sub_exp()`.

**Comment by G-Lauz** [comment]
Following the idea from an offline discussion:

The problems seem to be more about the organization of the data instead of a lack of features. We could take inspiration from an SQL database and use relationships between different entities:

### Fundamental entities:
- **dataset**: The data fed to the experiment
- **experiment**: The experiment itself, which is made of its source code and its hyperparameters
- **result**: The outputs of the experiments, the metrics of performance, the artifact (plot, image, etc.) and the model (if learning something like neural network does)

Each of those entities are stored in a registry of entities and are identified by a hash that could be used as a key to aggregate them under the same "run". A "run" could look something like:

```JSON
{
    "git": "SHA-1-OF-THE-LAST-COMMIT",
    "project": "my_project_name",
    "run-name": "enlightened_white-eye",
    "date": "2025-07-03T20:08:14.071247Z",
    "dataset": "SHA-1-OF-THE-DATASET",
    "experience": "SHA-1-OF-THE-EXPERIENCE",
    "result": "SHA-1-OF-THE-RESULTS"
}
```

### The "sub-experiment" feature

By dissociating the entities, we could implement any logic based on the relations between runs to fetch the required informations. As example, a "sub-experiment" features could be implemented as:

```JSON
{
    "git": "SHA-1-OF-THE-LAST-COMMIT",
    "project": "my_project_name",
    "run-name": "enlightened_white-eye",
    "date": "2025-07-03T20:08:14.071247Z",
    "sub-runs": [
        {
            "dataset": "SHA-1-OF-THE-DATASET-1",
            "experience": "SHA-1-OF-THE-EXPERIENCE-1",
            "result": "SHA-1-OF-THE-RESULTS-1"
        },
        {
            "dataset": "SHA-1-OF-THE-DATASET-2",
            "experience": "SHA-1-OF-THE-EXPERIENCE-2",
            "result": "SHA-1-OF-THE-RESULTS-2"
        },
    ]
}
```

### Other side effects

This pattern also open up the possibility to scale the application to a database easily instead of staying local.

This could also extend the feature of the experiments version control to datasets version control and also results version control which could be interesting as working with AI most of the time start with the quality of the data.

**Comment by Andlat** [comment]
I like that 👀 We'll have to think about how to integrate the dataset. Would you consider training vs evaluation two different experiments ? The dataset could be common between them, or separate

The can see the DB working with either SQL (sqlite ?)  or NoSQL (redis maybe ?)
I feel like nosql might be an easier setup, but sql is maybe more appropriate since we have static fields
